src/realm/array_mixed.cpp

```diff
@@ -0,0 +1,348 @@
+/*************************************************************************
+ *
+ * Copyright 2018 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#include <realm/array_mixed.hpp>
+#include <realm/array_basic.hpp>
+
+using namespace realm;
+
+ArrayMixed::ArrayMixed(Allocator& a)
+    : Array(a)
+    , m_composite(a)
+    , m_ints(a)
+    , m_int_pairs(a)
+    , m_strings(a)
+{
+    m_composite.set_parent(this, s_primary_index);
+}
+
+void ArrayMixed::create()
+{
+    Array::create(type_HasRefs, false, s_root_size);
+    m_composite.create(type_Normal);
+    m_composite.update_parent();
+}
+
+void ArrayMixed::init_from_mem(MemRef mem) noexcept
+{
+    Array::init_from_mem(mem);
+    m_composite.init_from_parent();
+}
+
+void ArrayMixed::add(Mixed value)
+{
+    if (value.is_null()) {
+        m_composite.add(0);
+        return;
+    }
+    m_composite.add(store(value));
+}
+
+
+void ArrayMixed::set(size_t ndx, Mixed value)
+{
+    if (value.is_null()) {
+        set_null(ndx);
+        return;
+    }
+    erase_linked_payload(ndx);
+    m_composite.set(ndx, store(value));
+}
+
+void ArrayMixed::insert(size_t ndx, Mixed value)
+{
+    if (value.is_null()) {
+        m_composite.insert(ndx, 0);
+        return;
+    }
+    m_composite.insert(ndx, store(value));
+}
+
+void ArrayMixed::set_null(size_t ndx)
+{
+    auto val = m_composite.get(ndx);
+    if (val) {
+        erase_linked_payload(ndx);
+        m_composite.set(ndx, 0);
+    }
+}
+
+Mixed ArrayMixed::get(size_t ndx) const
+{
+    int64_t val = m_composite.get(ndx);
+    int64_t int_val = val >> s_data_shift;
+    size_t payload_ndx = size_t(int_val);
+    DataType type = DataType((val & s_data_type_mask) - 1);
```

Comment:
You need to handle getting a null value here. What it looks like you had in mind was to use the first bit of the type to encode if the type was nullable or not? This seems like a good idea, as when we get a null value, we need to return null of the correct type. This means that `set_null` cannot set 0 in the composite, as it still needs to encode the type.

Reply:
We discussed this in the office and concluded that the existing implementation is correct; excepting that there should be an early out here if the value is null. There is no concept of a typed null in a Mixed value.

Reply:
Fixed